### PR TITLE
add deepEP training

### PR DESCRIPTION
**Pull request by faresobeid**
> [!NOTE]
> **High Risk**
> High risk because it changes core distributed training behavior (MoE expert-parallel communication, checkpointing boundaries, and fused loss/LM head code paths) and adds new CUDA-kernel-backed dependencies that can fail at runtime on unsupported hardware or configs.
> 
> **Overview**
> Adds a new MoE expert-parallel communication backend, `model.ep_comm_backend='deepep'`, wiring DeepEP dispatch/combine into `MoE`/`LatentMoE` and model setup with runtime validation plus new tuning knobs (`model.deepep_num_sms`, `model.deepep_token_chunk_size`).
> 
> Introduces **selective activation checkpointing** via `model.ac.mode`/`model.ac.targets`, with model/runtime validation and layer-method checkpoint patching to checkpoint only chosen subcomponents (with fallback to full checkpointing where unsupported).
> 
> Extends SFT with `loss_impl='quack_fused'` (quack-kernels chunked linear+CE) and adds optional quack-accelerated `RMSNorm` on CUDA; also updates env server orchestration/logging (worker auto-scaling, log directories, verifiers API move), removes trainer `model.tp` usage, adjusts parallel-dims calculations accordingly, and hardens checkpoint loads (`torch.load(..., weights_only=False)`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 43224f80bf44f946f2d92a5950e0a81e43423554. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: skills/toml-config/SKILL.md [hunk 1]
```diff
@@ -60,6 +60,29 @@ Use the string `"None"` in TOML to set a field to None:
...
+[model]
+ep = 8
+ep_comm_backend = "deepep"  # standard | deepep
+```
+
+CLI equivalent:
+
+```bash
+uv run trainer @ trainer.toml --model.ep 8 --model.ep-comm-backend deepep
+```
```
Comment (samsja) [inline skills/toml-config/SKILL.md]: lets remove this

File: skills/toml-config/SKILL.md [hunk 2]
```diff
@@ -60,6 +60,29 @@ Use the string `"None"` in TOML to set a field to None:
...
+
+Important constraints:
+
+- `model.ep_comm_backend != "standard"` requires `model.ep > 1`
+- keep `model.tp = 1` for the current DeepEP path
+- DeepEP is not currently supported together with trainer LoRA
+- selective AC target `routed_experts` is supported and checkpoints only the local expert matmuls
+
```
Comment (samsja) [inline skills/toml-config/SKILL.md]: lets remove this

File: src/prime_rl/configs/trainer.py [hunk 3]
```diff
@@ -363,6 +385,24 @@ def flash_attention_4_only_with_custom_impl(self):
...
+    def validate_ep_comm_backend(self):
+        if self.ep_comm_backend == "standard":
+            return self
+
+        if self.ep <= 1:
+            raise ValueError(f"model.ep_comm_backend='{self.ep_comm_backend}' requires model.ep > 1.")
+        if self.tp != 1:
+            raise ValueError(
+                f"model.ep_comm_backend='{self.ep_comm_backend}' is currently only supported with model.tp=1."
+            )
```
Comment (samsja) [inline src/prime_rl/configs/trainer.py]: we don;t have to support anyway

File: skills/installation/SKILL.md [hunk 4]
```diff
@@ -33,6 +33,44 @@ uv sync --group fp8-inference
...
+```bash
+NVSHMEM_DIR=$(
+  uv run python - <<'PY'
+import importlib.util
+spec = importlib.util.find_spec("nvidia.nvshmem")
+if not spec or not spec.submodule_search_locations:
+    raise SystemExit("nvidia.nvshmem not found in .venv")
+print(spec.submodule_search_locations[0])
+PY
+)
```
Comment (samsja) [inline skills/installation/SKILL.md]: would need ot find a way to unstakl this in the pyporject toml tho

File: src/prime_rl/trainer/distributed/deepep.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
+from dataclasses import dataclass
+
+import torch
+from torch.distributed import ProcessGroup
+
+try:
+    from deep_ep import Buffer
+    from deep_ep.utils import EventHandle, EventOverlap
+except ImportError as e:
+    raise ImportError("DeepEP is required for this backend. Install from https://github.com/deepseek-ai/DeepEP.") from e
```
Comment (samsja) [inline src/prime_rl/trainer/distributed/deepep.py]: would remove the try expect for import

File: src/prime_rl/trainer/distributed/deepep.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+
+_lib = torch.library.Library("deepep", "DEF")
+_lib.define(
+    "dispatch(Tensor x, Tensor topk_idx, Tensor topk_weights, "
+    "Tensor num_tokens_per_rank, Tensor num_tokens_per_rdma_rank, "
+    "Tensor is_token_in_rank, Tensor num_tokens_per_expert) "
+    "-> (Tensor, Tensor, Tensor, Tensor, Tensor)"
+)
+_lib.define("combine(Tensor x, Tensor handle_id) -> Tensor")
```
Comment (samsja) [inline src/prime_rl/trainer/distributed/deepep.py]: can we move all thus logic to a function that we call at the begining of the script ?

File: src/prime_rl/trainer/models/layers/lora/multi_moe.py [hunk 7]
```diff
@@ -309,6 +309,9 @@ def state_dict_for_adapter(self, idx: int) -> dict[str, torch.Tensor]:
         return state_dict
 
     def forward(self, x: torch.Tensor, num_tokens_per_expert: torch.Tensor) -> torch.Tensor:
+        if getattr(self.base_layer, "ep_comm_backend", "standard") != "standard":
+            raise ValueError("MoE LoRA adapters are not currently supported with the DeepEP backend.")
+
```
Comment (Jackmin801) [inline src/prime_rl/trainer/models/layers/lora/multi_moe.py]: whats the issue that occurs with lora enabled?

File: src/prime_rl/trainer/models/layers/moe.py [hunk 8]
```diff
@@ -434,6 +501,28 @@ def forward(
...
+            routed_output = self._run_routed_experts(hidden_states, num_tokens_per_expert)
+            routed_output = combine_tokens(routed_output, dispatch_state)
+
+            shared_output = self.shared_expert(x) if self.shared_expert is not None else None
+            sync_combine()
+
+            if shared_output is None:
+                return routed_output.reshape(bs, slen, dim)
+            return (shared_output + routed_output).reshape(bs, slen, dim)
+
```
Comment (Jackmin801) [inline src/prime_rl/trainer/models/layers/moe.py:567]: Can we put these as input and output hooks on the ParallelStyle rather than inline in the module like this?
